python/cudf/cudf/_lib/csv.pyx

```diff
@@ -207,10 +188,34 @@ cpdef read_csv(
     if prefix is not None:
         args.prefix = prefix.encode()
 
-    cdef unique_ptr[cpp_csv.reader] reader
-    with nogil:
-        reader = unique_ptr[cpp_csv.reader](new cpp_csv.reader(args))
+    # Create reader from source
+    cdef unique_ptr[csv_reader] reader
+    cdef const unsigned char[::1] buffer = None
+    if isinstance(filepath_or_buffer, BytesIO):
+        buffer = filepath_or_buffer.getbuffer()
+    elif isinstance(filepath_or_buffer, StringIO):
+        buffer = filepath_or_buffer.read().encode()
+    elif isinstance(filepath_or_buffer, bytes):
+        buffer = filepath_or_buffer
+
+    if buffer is None:
+        if not os.path.isfile(filepath_or_buffer):
+            raise FileNotFoundError(
+                errno.ENOENT, os.strerror(errno.ENOENT), filepath_or_buffer
+            )
+        reader = unique_ptr[csv_reader](
+            new csv_reader(str(filepath_or_buffer).encode(), args)
+        )
+    elif buffer.shape[0] != 0:
+        reader = unique_ptr[csv_reader](
+            new csv_reader(<char *>&buffer[0], buffer.shape[0], args)
+        )
+    else:
+        reader = unique_ptr[csv_reader](
+            new csv_reader(<char *>NULL, 0, args)
+        )
```

Comment:
You're no longer releasing the GIL here which you should be, the calls to the `csv_reader` should be in a `with nogil` context

Reply:
Done, along with the read() calls as well.